machida3/machida.pony

```diff
@@ -382,12 +380,12 @@ class PyTCPEncoder is TCPSinkEncoder[PyData val]
   fun apply(data: PyData val, wb: Writer): Array[ByteSeq] val =>
     let byte_buffer = Machida.sink_encoder_encode(_sink_encoder, data.obj())
     if not Machida.is_py_none(byte_buffer) then
-      let byte_string = @PyBytes_AsString(byte_buffer)
+      let byte_string = @py_bytes_or_unicode_as_char(byte_buffer)
 
       if not byte_string.is_null() then
         let arr = recover val
           // create a temporary Array[U8] wrapper for the C array, then clone it
-          Array[U8].from_cpointer(@PyBytes_AsString(byte_buffer),
+          Array[U8].from_cpointer(@py_bytes_or_unicode_as_char(byte_buffer),
```

Comment:
This implies that the user should be able to use bytes or unicode values as the output of an encoder. I think we should restrict encoders to returning bytes, not strings. Is the intention to allow unicode strings or bytes here? If so, why?

Unless there's a good reason for allowing unicode strings AND bytes in a lot of places, I think we should break `py_bytes_or_unicode_as_char` into two functions, `py_bytes_as_char` and `py_unicode_as_char`, and then use those functions in the appropriate places where we expect either bytes or unicode. Having this catch-all function makes me nervous and allowing bytes and unicode strings in so many places seems unnecessarily permissive and potentially error-prone.

Reply:
while the user _should_ return bytes from that function, I think we should catch and handle that case in the python api side, and produce a useful error there. 
In machida itself, I think segfaulting because we got the wrong kind of string is worse than simply sending it out as bytes, regardless of what it is.
We could put a special check for it being bytes here, but again, I think it's easier to produce useful error messages (and pointing at the offending code) from the python api side.

What do you think?

Reply:
I agree, we shouldn't segfault. What I'm saying is that I think if we're already enforcing type constraints on the Python side then we should be able to know that we could call `py_bytes_as_char` or `py_unicode_as_char` instead of having to rely on a single function that handles both cases.

Reply:
In principle I agree that we should be strict with this, something I experienced working on these parts was that the error messaging form Machida was a lot less than what's necessary for a good developer experience. As I'm not aware of any spec for what the application developer can expect, maybe a handful of user stories would be enough guidance for what this needs to be.

Reply:
The guidance here is that anything that would be expected to be a string (such as names) should be a string, anything that is a key or a data blob (for example: the return value of a partition function, a serialized representation of an object, or the output of a tcp encoder) should be bytes. @nisanharamati @caj-larsson does that sound like a reasonable description of the expected behavior?

Reply:
I disagree. 
I think that in most places, machida should tolerate any text type that can be reliably compared as bytes (e.g. bytestrings or unicode). Most of these values are used for either of 2 things:

1. uniquely identify a thing
2. print some text to stdout

Both bytestrings and unicode work fine here, so I don't understand the benefit in restricting users to only one or the other. We can easily distinguish between the two types on the pony/c side, and adding this requirement makes user code less friendly because of both the potential errors, and the potential for user values to be either type (for example if you read them from a file or a buffer with a 3rd party parser that may return either type).
 
I also think there's some performance impact from requiring code in the hot path to also execute `encode()` on a string (e.g. partition function), but I don't know if it's significant enough to be of concern:
```
$ python3 -m timeit -s "x = 'hello'*1000" "x.encode()"
10000000 loops, best of 3: 0.149 usec per loop
$ python3 -m timeit -s "x = 'hello'*10000" "x.encode()"
1000000 loops, best of 3: 1.22 usec per loop
```
Maybe for some cases.

The places where I agree that machida should only accept byestrings are:
- decoder input 
- encoder output
- serialization output
- deserialization input


For the serialization/deserialization case I think the main limitation arises because we don't know at deserialization time what the original serialization output type was, so we can choose whether to create a unicode or a bytestring python object as the argument for the user's deserialization method.
I think that's a tractable problem, but maybe it's outside the scope of this PR?

Reply:
As a Python programmer I like @nisanharamati expect the system to have some leniency between these types but as a C programmer I see them as very different things and that the interface should be strict. Is this an issue best solved in the python library?